paymentgateway/paypal/classes/paymentgateway.php

```diff
@@ -0,0 +1,197 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Class for a specific payment gateway object.
+ *
+ * @package     paymentgateway_paypal
+ * @author      Haruki Nakagawa
+ *
+ * @copyright   MAHQ
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ **/
+
+namespace paymentgateway_paypal;
+
+use moodle_exception;
+use context_system;
+
+defined ('MOODLE_INTERNAL') || die();
+
+class paymentgateway extends \tool_paymentplugin\paymentgateway\object_paymentgateway {
+
+    /**
+     * Alerts site admin of potential problems.
+     *
+     * @param string   $subject email subject
+     * @param stdClass $data    PayPal IPN data
+     */
+    private function message_paypal_error_to_admin($subject, $data) {
+        global $PAGE;
+        $PAGE->set_context(context_system::instance());
+
+        $admin = get_admin();
+        $site = get_site();
+
+        $message = "$site->fullname:  Transaction failed.\n\n$subject\n\n";
+
+        foreach ($data as $key => $value) {
+            $message .= "$key => $value\n";
+        }
+
+        $eventdata = new \core\message\message();
+        $eventdata->courseid          = empty($data->courseid) ? SITEID : $data->courseid;
+        $eventdata->modulename        = 'moodle';
+        $eventdata->component         = 'paymentgateway_paypal';
+        $eventdata->name              = 'payment_paypal_error';
+        $eventdata->userfrom          = $admin;
+        $eventdata->userto            = $admin;
+        $eventdata->subject           = "PAYPAL ERROR: ".$subject;
+        $eventdata->fullmessage       = $message;
+        $eventdata->fullmessageformat = FORMAT_PLAIN;
```

Comment:
if you make this into format_moodle, you get some extra goodies for free, like link filtering making hyperlinks from urls